### DISCUSSION TITLE
Memory issues on Firedrake

### DISCUSSION THREAD

**Discussion by miguelcoolchips** [firedrakeproject/firedrake#2865] [Firedrake support] [answered]
I have recently learned about some of the memory issues with Firedrake. I think it would be a good idea to list them all somewhere on the website. The ones I am aware of are:
- All `Constant` objects need the `mesh` to be passed in their argument `domain`.
- Cache cleaning for long-running simulations [(details)](https://github.com/OP2/PyOP2/issues/693)

Are there other issues I should be aware of? In general, what is the procedure to follow to detect memory leaks? How to interpret the output of `petsc4py.PETSc.garbage_view()`?

**Comment by connorjward** [comment, marked as answer]
In general we try to hide memory concerns from the users as much as possible. I think that the two issues you mention here should be regarded as bugs and hence fixed, rather than documented. Specifically:

- We have already figured out a solution to the `Constant` needing a `domain` issue and hopefully will have it fixed within the month.
- For the cache cleaning issue I think I need to implement an LRU cache to bound the memory usage. This is slightly complicated by the need to make sure it is consistent across a communicator but I don't think it would be too hard.

To detect memory leaks we recommend using `memory_profiler`. This is documented in the manual [here](https://firedrakeproject.org/optimising.html#memory-profiler). We have also recently documented petsc4py so, once merged, questions about `garbage_view` should hopefully be resolved.

@JDBetteridge may have more to say on this.
